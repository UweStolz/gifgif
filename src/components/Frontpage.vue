<template>
  <v-container
    fluid
    fill-height
  >
    <Snackbar
      :visibility.sync="showSnackbar"
      :message="snackbarMessage"
    />
    <v-row
      align="start"
      justify="center"
      no-gutters
    >
      <v-card
        flat
        tile
        height="100%"
        width="100%"
      >
        <v-col cols="auto">
          <v-carousel
            v-model="carouselModel"
            hide-delimiters
            height="35vmax"
            :show-arrows="!$vuetify.breakpoint.xsOnly || gifsList.length > 1"
            :show-arrows-on-hover="!$vuetify.breakpoint.xsOnly || gifsList.length > 1"
            @change="updateCarouselModel"
          >
            <v-carousel-item
              v-for="(gif, i) in gifsList"
              :key="i"
            >
              <v-img
                :src="gif.url"
                :lazy-src="gif.url"
                contain
                height="100%"
                width="100%"
              >
                <template v-slot:placeholder>
                  <linear-progress />
                </template>
              </v-img>
            </v-carousel-item>
          </v-carousel>
        </v-col>
      </v-card>
    </v-row>
    <v-row
      align="start"
      justify="center"
      no-gutters
    >
      <v-col cols="auto">
        <card-actions
          :rating.sync="rating"
          :show-rating="true"
          :is-frontpage="true"
          :sync-function="getGifList"
          :image-id.sync="currentId"
          :image-data.sync="imageData"
          :preview-image-data.sync="previewData"
        />
      </v-col>
    </v-row>
  </v-container>
</template>

<script lang="ts">
import { Vue, Component } from 'vue-property-decorator';
import CardActions from '@/components/shared/CardActions.vue';
import LinearProgress from '@/components/shared/LinearProgress.vue';
import Snackbar from '@/components/shared/Snackbar.vue';
import {
  getTrendingGifsListFromGiphy,
  getRandomGifFromGiphy,
} from '../request';

@Component({
  components: {
    CardActions,
    LinearProgress,
    Snackbar,
  },
})

export default class Frontpage extends Vue {
  async getGifList(): Promise<void> {
    let yieldedError = false;
    if (this.$store.state.gifMode === 'trending') {
      const listFromGiphy: Giphy.Response = await getTrendingGifsListFromGiphy();
      if (listFromGiphy.length) { this.buildGifList(listFromGiphy); } else { yieldedError = true; }
    } else {
      const randomGifObject: Giphy.GIFObject = await getRandomGifFromGiphy();
      if (randomGifObject.id) {
        const itemObject = [
          {
            id: randomGifObject.id,
            url: randomGifObject.images.original.webp,
            previewUrl: randomGifObject.images.fixed_width.webp,
          },
        ];
        this.gifsList = itemObject;
      } else { yieldedError = true; }
    }
    if (!yieldedError) { await this.updateCarouselModel(); }
  }

  showSnackbar: boolean = false;

  snackbarMessage: string = 'gifgif was updated, please reload :)';

  imageKey: number = 0

  rating: number = 0;

  imageData: Blob | string = '';

  previewData: Blob | string = '';

  gifsList: BuiltGifLists = [];

  carouselModel: number = 0;

  currentId: string = '';

  async mounted(): Promise<void> {
    this.$store.watch(
      () => this.$store.state.gifMode,
      async () => {
        await this.getGifList();
      },
    );
    this.$store.watch(
      () => this.$store.state.pwaUpdated,
      () => {
        this.showSnackbar = true;
      },
    );
    await this.getGifList();
  }

  async updateCarouselModel(index: number = 0): Promise<void> {
    this.carouselModel = index;
    await this.getRating();
  }

  buildGifList(listFromGiphy: Giphy.Response): void {
    const builtGifList: BuiltGifLists = [];
    listFromGiphy.forEach((item) => {
      const itemObject: BuiltGifListObject = {
        id: item.id,
        url: item.images.original.webp.length ? item.images.original.webp : item.images.original.url,
        previewUrl: item.images.fixed_width.webp,
      };
      builtGifList.push(itemObject);
    });
    this.gifsList = builtGifList;
  }

  async setGifData() {
    this.currentId = this.gifsList[this.carouselModel].id;
    this.imageData = this.gifsList[this.carouselModel].url;
    this.previewData = this.gifsList[this.carouselModel].previewUrl;
  }

  async getRating() {
    await this.setGifData();
    const gifData: Database.GifData | undefined = await this.$store.dispatch('getGifData', `ggid-${this.currentId}`);
    this.rating = gifData?.rating ? this.rating = gifData.rating : this.rating = 0;
  }
}
</script>
