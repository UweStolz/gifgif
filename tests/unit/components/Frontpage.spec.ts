import Frontpage from '@/components/Frontpage.vue';
import shallow, { store, fetchMock } from '../../helper';

describe('Frontpage', () => {
  it('renders properly', async () => {
    const mockResponse = {
      data:
      {
        id: 'id-123',
        images: {
          original: {
            webp: 'http://image.webp',
          },
          fixed_width: {
            webp: 'http://preview.webp',
          },
        },
      },
    };
    fetchMock.resetMocks();
    fetchMock.mockResponse(JSON.stringify(mockResponse));
    const wrapper = shallow(Frontpage);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('Do not update if random GIF yields an error', async () => {
    store.state.gifMode = 'random';
    const mockResponse = {
      data:
      {
        id: '',
        images: {
          original: {
            webp: 'http://image.webp',
          },
          fixed_width: {
            webp: 'http://preview.webp',
          },
        },
      },
    };
    fetchMock.resetMocks();
    fetchMock.mockResponse(JSON.stringify(mockResponse));
    const wrapper = shallow(Frontpage);
    await wrapper.vm.$nextTick();
    await new Promise((resolve) => setTimeout(resolve));
    expect(wrapper).not.toHaveDispatched('getGifData');
  });

  it('Get trending GIFs on mount', async () => {
    store.state.gifMode = 'trending';
    const mockResponse = {
      data: [
        {
          id: 'id-123',
          images: {
            original: {
              webp: 'http://image.webp',
            },
            fixed_width: {
              webp: 'http://preview.webp',
            },
          },
        },
      ],
    };
    fetchMock.resetMocks();
    fetchMock.mockResponse(JSON.stringify(mockResponse));
    const wrapper = shallow(Frontpage);
    await wrapper.vm.$nextTick();
    await new Promise((resolve) => setTimeout(resolve));
    const mockRandomGifObj = [{ id: 'id-123', previewUrl: 'http://preview.webp', url: 'http://image.webp' }];
    expect(wrapper.vm.$data.gifsList).toStrictEqual(mockRandomGifObj);
    expect(wrapper).toHaveDispatched('getGifData');
  });

  it('Get trending GIFs on mount, with fallback', async () => {
    store.state.gifMode = 'trending';
    const mockResponse = {
      data: [
        {
          id: 'id-123',
          images: {
            original: {
              webp: '',
              url: 'http://image.gif',
            },
            fixed_width: {
              webp: 'http://preview.webp',
            },
          },
        },
      ],
    };
    fetchMock.resetMocks();
    fetchMock.mockResponse(JSON.stringify(mockResponse));
    const wrapper = shallow(Frontpage);
    await wrapper.vm.$nextTick();
    await new Promise((resolve) => setTimeout(resolve));
    const mockRandomGifObj = [{ id: 'id-123', previewUrl: 'http://preview.webp', url: 'http://image.gif' }];
    expect(wrapper.vm.$data.gifsList).toStrictEqual(mockRandomGifObj);
    expect(wrapper).toHaveDispatched('getGifData');
  });

  it('Get random GIF on mount', async () => {
    store.state.gifMode = 'random';
    const mockResponse = {
      data:
      {
        id: 'id-123',
        images: {
          original: {
            webp: 'http://image.webp',
          },
          fixed_width: {
            webp: 'http://preview.webp',
          },
        },
      },
    };
    fetchMock.resetMocks();
    fetchMock.mockResponse(JSON.stringify(mockResponse));
    const wrapper = shallow(Frontpage);
    await wrapper.vm.$nextTick();
    await new Promise((resolve) => setTimeout(resolve));
    const mockRandomGifObj = [{ id: 'id-123', previewUrl: 'http://preview.webp', url: 'http://image.webp' }];
    expect(wrapper.vm.$data.gifsList).toStrictEqual(mockRandomGifObj);
    expect(wrapper).toHaveDispatched('getGifData');
  });
});
