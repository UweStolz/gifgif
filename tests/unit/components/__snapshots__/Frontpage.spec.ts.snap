// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Frontpage.vue renders properly 1`] = `
<v-container-stub
  fill-height=""
  fluid="true"
  tag="div"
>
  <v-row-stub
    align="start"
    justify="center"
    nogutters="true"
    tag="div"
  >
    <v-card-stub
      flat="true"
      height="100%"
      loaderheight="4"
      tag="div"
      tile="true"
      width="100%"
    >
      <v-col-stub
        cols="auto"
        tag="div"
      >
        <v-carousel-stub
          activeclass="v-window-item--active"
          continuous="true"
          delimitericon="$delimiter"
          height="30vmax"
          hidedelimiters="true"
          interval="6000"
          mandatory="true"
          nexticon="$next"
          previcon="$prev"
          value="0"
        />
      </v-col-stub>
    </v-card-stub>
  </v-row-stub>
   
  <v-row-stub
    align="start"
    justify="center"
    nogutters="true"
    tag="div"
  >
    <v-col-stub
      cols="auto"
      tag="div"
    >
      <card-actions-stub
        imagedata=""
        imageid=""
        isfrontpage="true"
        previewimagedata=""
        rating="0"
        showrating="true"
        syncfunction="function () { [native code] }"
      />
    </v-col-stub>
  </v-row-stub>
</v-container-stub>
`;
